nemo_aligner/algorithms/ppo.py

```diff
@@ -201,84 +270,121 @@ def post_process_tensor(tensor):
+        if compute_init_policy_kl:
+            self.timer.start("init_logprobs")
+            rollout_init_logprobs = self.model.get_init_policy_logprobs(batched_response_tokens)
+            balanced_local_batch["init_logprobs"] = rollout_init_logprobs
+            timer_metrics["init_logprobs"] = self.timer.stop_and_get_time("init_logprobs")
+
+        # we send the request in sharded context, so we need to keep this sharding and then undo it
+        with reshard_context():
+            self.timer.start("critic_wait")
+            rm_value_rollout_batches = []
+            for future in futures:
+                rewards, values = future.result() if isinstance(future, FutureResult) else future
+                rm_value_rollout_batches.append({"rewards": rewards, "values": values})
+            timer_metrics["critic_wait"] = self.timer.stop_and_get_time("critic_wait")
+
+            unbalanced_rm_value_batch = PPORolloutBatch.from_rollout_batches(
+                rm_value_rollout_batches,
+                eos_id=self.model.tokenizer.eos_id,
+                rollout_batch_seq_length=padded_rollout_sequence_length,
+            )
+            global_rm_value_batch = unbalanced_rm_value_batch.gather_and_balance_globally()
 
-        tensor_to_accumulate = torch.tensor(
-            [metrics["response_lengths"], metrics["prompt_lengths"], metrics["rewards"], num_samples],
-            dtype=torch.float32,
-            device=torch.cuda.current_device(),
+        # chunking needs to be outside of reshard region
+        # NOTE: the seed here must be the same as the chunk before since we need to shuffle
+        # these values the same way as the other values
+        balanced_rm_value_batch = global_rm_value_batch.chunk(
+            parallel_state.get_data_parallel_rank(), parallel_state.get_data_parallel_world_size(), self.step
         )
-        torch.distributed.all_reduce(tensor_to_accumulate, group=parallel_state.get_data_parallel_group())
+        balanced_local_batch.update(balanced_rm_value_batch)
+
+        global_rollout_batch.update(global_rm_value_batch)
+
+        return balanced_local_batch, cpu_dict(self.compute_rollout_metrics(global_rollout_batch)), timer_metrics
+
+    def compute_rollout_metrics(self, rollout_batch):
+        table = {}
 
-        (
-            global_response_lengths,
-            global_prompt_lengths,
-            global_rewards,
-            global_num_samples,
-        ) = tensor_to_accumulate.tolist()
+        prompt_lengths = rollout_batch["prompt_lengths"]
+        response_lengths = rollout_batch["response_lengths"]
+        response_tokens = rollout_batch["response_tokens"]
+        rewards = rollout_batch["rewards"]
+        is_end = rollout_batch["is_end"]
+
+        reward = rewards[0]
+        prompt_length = prompt_lengths[0]
+        response_length = response_lengths[0]
+        response_token = response_tokens[0]
```

Comment:
Could use a comment as to why we're accessing `[0]` everywhere here

Reply:
424383b78853b6ca101102400b22db7768b14e93